web/pandas/pdeps/0007-copy-on-write.md

```diff
@@ -0,0 +1,575 @@
+probably be too noisy. Indexing is too common an operation to include a warning (even if
+we limit it to just those operations that previously returned views). However, we can
+already implement this proposal and make it available before becoming the default
+behaviour, such that users can opt-in and test their code (this is currently possible
+with `pd.options.mode.copy_on_write = True`). Further, it should be possible to also
+provide a "warning" mode that raises warnings for all cases that will change behaviour
+under the Copy-on-Write proposal. If we have such a warning mode, we could also enable
+those warnings by default in a last release before the major version bump. We can then
+provide a clearly documented upgrade path to first enable the warnings, fix all
+warnings, and then enable the Copy-on-Write mode and ensure your code is still working,
+and then finally upgrade to the new major release.
+
+## Implementation
+
+An initial implementation has been
+[merged](https://github.com/pandas-dev/pandas/pull/46958) and is available since pandas
+1.5 (and significantly improved in 2.0). It uses weakrefs to keep track of whether the
+data of a Dataframe/Series are viewing the data of another (pandas) object or are being
+viewed by another object. This way, whenever the series/dataframe gets modified, we can
+check if its data first needs to be copied before mutating it.
+
+To test the implementation and experiment with the new behaviour, you can can
+enable it with the following option:
+
+```python
+>>> pd.options.mode.copy_on_write = True
+```
+
+after importing pandas (or setting the `PANDAS_COPY_ON_WRITE=1` environment variable
+before importing pandas).
+
+## Concrete examples
+
+### Chained assignment
+
+Consider a "classic" case of chained indexing, which was the original motivation for the SettingWithCopy warning:
+
+```python
+>>> df[df['B'] > 4]['B'] = 10
+```
+
+That is roughly equivalent to
+
+```python
+>>> df2 = df[df['B'] > 4]  # Copy under NumPy's rules
+>>> df2['B'] = 10  # Update (the copy) df2, df not changed
+>>> del df2  # All references to df2 are lost, goes out of scope
+```
+
+And so `df` is not modified. For this reason, the SettingWithCopyWarning was introduced.
+
+_With this proposal_, any result of an indexing operation behaves as a copy
+(Copy-on-Write), and thus chained assignment will _never_ work. Given that there is then
+no ambiguity, the idea is to drop the warning.
+
+The above example is a case where chained assignment doesn't work with current pandas.
+But there are of course also patterns with chained assignment that currently _do_ work
+and are used. _With this proposal_, any chained assignment will not work, and so those
+cases will stop working (e.g. the case above but switching the order):
+
+```python
+>>> df['B'][df['B'] > 4] = 10
+# or
+>>> df['B'][0:5] = 10
+```
+
+It might be possible to detect this and raise an error (since it can never do what the
```

Comment:
I think the proposal has to be more concrete here.  We can't say "it might".  The proposal should say "It will be possible", or "It is not possible", or "under certain circumstances, it is possible "

Reply:
We don’t know about this yet. We ran into a problem when doing a setitem from cython

Reply:
Updated with what we are doing right now